nixos/glance: init at liberty version

nlewo opened the pull request:
###### Motivation for this change


###### Things done

- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

nlewo committed:
nixos/glance: init at liberty version

This commit is based on initial works made by domenkozar.

nlewo commented:
cc @spinus @Mic92

edolstra commented on nixos/modules/virtualisation/openstack/glance.nix:
```diff
@@ -0,0 +1,249 @@
+          the first Glance execution by the systemd post start.
+
+	  The keystone admin account is used to create required
+	  Keystone resource for the Glance service.
```
Indentation is off. Also, you need to use `<para>` if you want the paragraph breaks to show up.

nlewo replied:
Done.

spinus commented:
@nlewo would be nice to add similar test to one you added for keystone, start some vm, run services, make sure they are working. Maybe even integration test for keystone would be good as well.

nlewo committed:
nixos/glance: add test

The test starts the glance service, creates a nixos image and ensures Glance
list it.

Note the test also starts the Keystone service since it is required
by Glance.

nlewo commented:
@edolstra sorry for indentation problems. I fixed them.
@spinus I made it but I forgot to git add the file :( I added it in a new commit.

spinus commented on nixos/modules/virtualisation/openstack/glance.nix:
```diff
@@ -0,0 +1,249 @@
+      type = types.str;
+      default = http://localhost:5000;
+      description = ''
+        Complete public Identity API endpoint. Note this is unversionned.
```
If this should point to keystone maybe worth to mention it. What public means here? Must be accessible from the internet? (why?)

nlewo replied:
Openstack uses three kind of API endpoints (public, internal, admin) which have different policies. For instance, on the public endpoint, even if you have the admin credentials, you would not be able to create a user account.

I think this means Glance doens't require internal or admin privileges to work. 
This has not to be accessible from Internet.

I'll mention Keystone in the description.

spinus commented on nixos/modules/virtualisation/openstack/glance.nix:
```diff
@@ -0,0 +1,249 @@
+    # Note: when changing the default, make it conditional on
+    # ‘system.stateVersion’ to maintain compatibility with existing
+    # systems!
+    virtualisation.openstack.glance.package = mkDefault pkgs.glance;
```
I think would be easier just to add "default" to mkOption above rather than modifying it here.

nlewo replied:
In fact, I don't understand why it is implemented like this (this is inhertited from @domenkozar works), it's why I let it like this...

domenkozar replied:
Default packages used to break manual, but that can be mitigated now with `defaultText` so we can change this.

nlewo replied:
@domenkozar thanks for your help but maybe i didn't understand your answer or it has been fixed... because in my last commit, I set the default package and in the manual the default package text is ` (build of glance-11.0.0) ` which seems to be clean and explicit.

spinus commented:
@nlewo in general looks ok (for testing)

nlewo committed:
nixos/glance: set default glance package

Before, it was overridden in the config section to avoid problem related
to manual generation.

Mic92 commented:
Thanks!
